### DISCUSSION TITLE
Can connectrpc work with go grpc middlewares?

### DISCUSSION THREAD

**Discussion by tuananhlai** [connectrpc/connect-go#686] [Q&A] [answered]
Hi everyone, I am wondering whether connectrpc server can work with existing Go grpc middlewares, like the one you find in [this repository](https://github.com/grpc-ecosystem/go-grpc-middleware). I didn't see any mentions of this interoperability in the docs, so I'd appreciate it if someone can confirm this information.

Thank you in advanced for your help.

**Comment by jhump** [comment, marked as answer]
@tuananhlai, no I'm afraid it does not. But it _does_, on the other hand work with regular "net/http" middleware. So many common functions, like authn|z, rate limiting, etc can still easily be integrated with Connect as there a "net/http"-based middleware libraries for these.

One thing that can be done to keep using the gRPC libraries, including this middleware, is to continue using the grpc-go module for your server implementation, but use [vanguard-go](https://github.com/connectrpc/vanguard-go) as middleware in front of that. That allows your gRPC server to accept traffic in the Connect protocol, so you can use Connect from web and mobile clients without having to migrate server code to the connect-go server library.

**Reply by tuananhlai** [reply]
Thank you for your reply. I'm just curious, but if the Connect server already works with HTTP middlewares, is there a particular reason why the Connect team is implementing their own middlewares (like authn-go and cors-go)?

**Reply by mattrobenolt** [reply]
The CORS and Auth packages are an HTTP level middleware. They are just a way to generalize some ecosystem patterns. The CORS package just is constants so you don't need to know which headers to set to which values, it's to provide the defaults. Auth is mostly generic, with a little bit of uniformity around handling the Connect protocol, but there's no strict reason either of these would be required. You can just fully implement either without those packages.
